### PR TITLE
Track inputs sent to wallet-internal recipients

### PR DESCRIPTION
**Pull request by nuttycom**
This adds trial decryption with the internal IVK and correctly tracks payments sent to the wallet-internal address in the wallet database.

### CODE REVIEW

File: zcash_client_backend/src/decrypt.rs [hunk 1]
```diff
@@ -15,6 +16,17 @@ use zcash_primitives::{
 
 use crate::keys::UnifiedFullViewingKey;
 
+/// An enumeration of the possible relationships a TXO can have to the wallet.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum TransferType {
```
Comment (str4d) [inline zcash_client_backend/src/decrypt.rs:21]: Needs to be mentioned in the changelog as an addition.

File: zcash_client_backend/src/decrypt.rs [hunk 2]
```diff
@@ -15,6 +16,17 @@ use zcash_primitives::{
...
+    /// The transfer was received on one of the wallet's external addresses.
+    Incoming,
+    /// The transfer was received on one of the wallet's internal-only addresses.
+    WalletInternal,
+    /// The transfer was decrypted using one of the wallet's outgoing viewing keys.
+    Outgoing,
```
Comment (str4d) [inline zcash_client_backend/src/decrypt.rs:27]: Either these cases need more documentation, or the enum as a whole needs more documentation. In particular, it is _almost never_ true that a wallet will detect an outgoing transaction by decrypting with an OVK; instead a change output will be decrypted with an IVK. There are potential edge cases here as well when a transaction is e.g. sending funds between two accounts in a wallet, or involves an explicit send to a wallet's external address as well as a wallet-external address, which cannot be represented as a single instance of this enum.

File: zcash_client_backend/src/decrypt.rs [hunk 3]
```diff
@@ -33,7 +45,7 @@ pub struct DecryptedOutput {
     /// this is a logical output of the transaction.
     ///
     /// [`OutgoingViewingKey`]: zcash_primitives::keys::OutgoingViewingKey
-    pub outgoing: bool,
+    pub transfer_type: TransferType,
```
Comment (str4d) [inline zcash_client_backend/src/decrypt.rs:48]: - This is a change to a public type. Document it in the changelog.
- Per my above comment, a single instance of `TransferType` is insufficient to fully describe the possible transaction types. This needs to either be a set or a bitflag, or `TransferType` needs additional cases.

File: zcash_client_backend/src/decrypt.rs [hunk 4]
```diff
@@ -49,27 +61,37 @@ pub fn decrypt_transaction<P: consensus::Parameters>(
...
+                        try_sapling_note_decryption(params, height, &ivk_external, output)
+                            .map(|ret| (ret, TransferType::Incoming))
+                            .or_else(|| {
+                                try_sapling_note_decryption(params, height, &ivk_internal, output)
+                                    .map(|ret| (ret, TransferType::WalletInternal))
+                            })
+                            .or_else(|| {
+                                try_sapling_output_recovery(params, height, &ovk, output)
+                                    .map(|ret| (ret, TransferType::Outgoing))
+                            });
```
Comment (str4d) [inline zcash_client_backend/src/decrypt.rs]: This looks vaguely like some similar code in `zcashd`. We should document the logic here, and try to keep it in line with what we do in `zcashd`.

File: zcash_client_sqlite/src/wallet.rs [hunk 5]
```diff
@@ -1232,26 +1231,21 @@ pub fn put_sent_utxo<'a, P: consensus::Parameters>(
...
-)]
-pub fn insert_sent_note<'a, P: consensus::Parameters>(
+pub(crate) fn insert_sent_note<'a, P: consensus::Parameters>(
     stmts: &mut DataConnStmtCache<'a, P>,
     tx_ref: i64,
     output_index: usize,
-    account: AccountId,
-    to: &PaymentAddress,
+    from_account: AccountId,
+    to: &Recipient,
```
Comment (str4d) [inline zcash_client_sqlite/src/wallet.rs:1239]: The type here is invalid, as it allows a transparent address to be passed in.

File: zcash_client_sqlite/src/wallet.rs [hunk 6]
```diff
@@ -1170,14 +1169,14 @@ pub fn put_sent_note<'a, P: consensus::Parameters>(
     tx_ref: i64,
     output_index: usize,
     account: AccountId,
-    to: &PaymentAddress,
+    to: &Recipient,
```
Comment (str4d) [inline zcash_client_sqlite/src/wallet.rs:1172]: - The type here is invalid, as this allows a transparent address to be passed in.
- This is a change to a public deprecated API, making the deprecation meaningless. We should not change the public deprecated API, and instead either map to the new API internally (as was done below for `put_sent_utxo`) or remove the deprecated API.

File: zcash_client_sqlite/src/prepared.rs [hunk 7]
```diff
@@ -500,78 +579,6 @@ impl<'a, P> DataConnStmtCache<'a, P> {
...
-    /// Inserts a sent note into the wallet database.
-    ///
-    /// `output_index` is the index within the transaction that contains the recipient output:
-    ///
-    /// - If `to` is a Sapling address, this is an index into the Sapling outputs of the
-    ///   transaction.
-    /// - If `to` is a transparent address, this is an index into the transparent outputs of
-    ///   the transaction.
-    #[allow(clippy::too_many_arguments)]
-    pub(crate) fn stmt_insert_sent_note(
```
Comment (str4d) [inline zcash_client_sqlite/src/prepared.rs:512]: Why were these moved? It makes review significantly harder (for me on my 13" laptop screen that can't show two copies of this PR side-by-side to check what is moved and what is altered without line wrapping).

File: zcash_client_sqlite/src/wallet/init.rs [hunk 8]
```diff
@@ -122,8 +122,13 @@ fn init_wallet_db_internal<P: consensus::Parameters + 'static>(
     seed: Option<SecretVec<u8>>,
     target_migration: Option<Uuid>,
 ) -> Result<(), MigratorError<WalletMigrationError>> {
+    // Turn off foreign keys, and ensure that table replacement/modification
+    // does not break views
     wdb.conn
-        .execute("PRAGMA foreign_keys = OFF", [])
+        .execute_batch(
+            "PRAGMA foreign_keys = OFF;
+                        PRAGMA legacy_alter_table = TRUE;",
```
Comment (str4d) [inline zcash_client_sqlite/src/wallet/init.rs:130]:
```suggestion
            "PRAGMA foreign_keys = OFF;
            PRAGMA legacy_alter_table = TRUE;",
```

I also get suspicious when flags named `legacy_*` are being enabled; I'd prefer some more documentation explaining why this is necessary, and why it won't break with some future version of SQLite.

File: zcash_client_sqlite/src/wallet/init/migrations/sent_notes_to_internal.rs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+impl RusqliteMigration for Migration {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &rusqlite::Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch("ALTER TABLE sent_notes ADD COLUMN to_account INTEGER;")?;
+
+        // `to_account` should be null for all migrated rows, since internal addresses
+        // have not been used for change or shielding prior to this migration.
```
Comment (str4d) [inline zcash_client_sqlite/src/wallet/init/migrations/sent_notes_to_internal.rs:48]: This comment is confusing. `to_account` is guaranteed to be null because we just added it, and we'd have hit a database error if the column already existed.

I also do not understand why we are altering the table to add a `to_account` column in the first place, given that we then immediately replace that table with a new one. We should instead check that the column does not exist in `sent_notes` (to replace the error that we'd get above) and then just use the new table creation.

File: zcash_client_sqlite/src/lib.rs [hunk 10]
```diff
@@ -611,35 +622,26 @@ impl<'a, P: consensus::Parameters> WalletWrite for DataConnStmtCache<'a, P> {
...
+                        wallet::insert_sent_utxo(
+                            up,
+                            tx_ref,
+                            output.output_index,
+                            sent_tx.account,
+                            addr,
+                            output.value,
+                        )?
+                    }
+                    shielded_recipient => wallet::insert_sent_note(
```
Comment (str4d) [inline zcash_client_sqlite/src/lib.rs:636]: This incorrectly resolves the above TODO and introduces a bug. `wallet::insert_sent_note` internally hard-codes Sapling as the pool, which made sense because it explicitly took in a `sapling::PaymentAddress`. This change coerces all shielded output types to Sapling, which will cause corruption and (in the case of mixed transactions) data loss.

The correct approach is provide what the TODO says: more information about the output index. We need to pass through the output pool type all thr way through from where `sent_tx.outputs` is populated. Or alternatively, map the UA to its Sapling case, preserving the approach we were taking before (deferring Orchard support to a later PR).

File: zcash_client_sqlite/src/prepared.rs [hunk 11]
```diff
@@ -331,6 +336,80 @@ impl<'a, P> DataConnStmtCache<'a, P> {
...
+    /// Inserts a sent note into the wallet database.
+    ///
+    /// `output_index` is the index within the transaction that contains the recipient output:
+    ///
+    /// - If `to` is a Sapling address, this is an index into the Sapling outputs of the
+    ///   transaction.
+    /// - If `to` is a transparent address, this is an index into the transparent outputs of
+    ///   the transaction.
+    /// - If `to` is an internal account, this is an index into the Sapling outputs of the
+    ///   transaction.
```
Comment (str4d) [inline zcash_client_sqlite/src/prepared.rs:348]: This does not mention what happens when `to` is a UA, which is indicative of the bug I identified above.

**Review by str4d — CHANGES_REQUESTED** [review]
Reviewed 52e20ca195614ac7dcb15bf71b2cc53be20cbd66.

File: zcash_client_sqlite/src/lib.rs (hunk 10 above)
Comment (nuttycom) [reply zcash_client_sqlite/src/lib.rs:636]: I don't think this is correct. We need to be tracking the entire unified address to which we were requested to send the transaction; otherwise we're throwing away that critical information and can't retrieve it in the wallet history. This change was intentional. Right now, of course, we can only send to the Sapling pool, so that's why I didn't propagate the pool information through, but we need to retain the full UA somewhere.

File: zcash_client_sqlite/src/wallet/init/migrations/sent_notes_to_internal.rs [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        "Adds an identifier for the account that received an internal note to the sent_notes table"
+    }
+}
+
+impl RusqliteMigration for Migration {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &rusqlite::Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch("ALTER TABLE sent_notes ADD COLUMN to_account INTEGER;")?;
+
```
Comment (nuttycom) [inline zcash_client_sqlite/src/wallet/init/migrations/sent_notes_to_internal.rs:46]:
```suggestion
```

File: zcash_client_sqlite/src/prepared.rs (hunk 7 above)
Comment (nuttycom) [reply zcash_client_sqlite/src/prepared.rs:512]: These were moved because they now require the `consensus::Parameters` constraint on `P` where previously they did not.

File: zcash_client_sqlite/src/wallet.rs (hunk 6 above)
Comment (nuttycom) [reply zcash_client_sqlite/src/wallet.rs:1172]: I guess that this was a copy-paste error in https://github.com/zcash/librustzcash/commit/cb6a99384059b0bacff56a0e1e4de498b0a610bd

Comment (nuttycom) [reply zcash_client_sqlite/src/wallet.rs:1172]: Oh, wait. No. It's that the `sent_notes` table is actually also being used to track UTXOs, and the intermediate APIs are just obscuring this fact: https://github.com/zcash/librustzcash/blob/main/zcash_client_sqlite/src/wallet.rs#L1230-L1288

So the correct thing, as of right now, is to get rid of basically all of this and instead maybe rename the `sent_notes` table to something that more accurately reflects its role, and I guess re-create the views so that they're showing the right information.

Comment (nuttycom) [reply zcash_client_sqlite/src/wallet.rs]: Fixed in https://github.com/zcash/librustzcash/pull/659/commits/179e22b76d8a34eff7cfc995683dcf34692816ac

File: zcash_client_sqlite/src/wallet.rs (hunk 5 above)
Comment (nuttycom) [reply zcash_client_sqlite/src/wallet.rs]: It's actually not; see https://github.com/zcash/librustzcash/pull/659/commits/179e22b76d8a34eff7cfc995683dcf34692816ac

File: zcash_client_sqlite/src/lib.rs (hunk 10 above)
Comment (nuttycom) [reply zcash_client_sqlite/src/lib.rs]: Fixed in https://github.com/zcash/librustzcash/pull/659/commits/179e22b76d8a34eff7cfc995683dcf34692816ac by propagating the pool information in from where it's definitively known.

File: zcash_client_sqlite/src/prepared.rs (hunk 11 above)
Comment (nuttycom) [reply zcash_client_sqlite/src/prepared.rs:351]: Not a bug.

File: zcash_client_backend/src/data_api.rs [hunk 13]
```diff
@@ -248,21 +248,27 @@ pub struct SentTransaction<'a> {
     pub tx: &'a Transaction,
     pub created: time::OffsetDateTime,
     pub account: AccountId,
-    pub outputs: Vec<SentTransactionOutput<'a>>,
+    pub outputs: Vec<SentTransactionOutput>,
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/data_api.rs:251]: The `SentTransactionOutput` is new as of this release, but it had been only incidentally mentioned in the `Changed` section of the changelog. Adding it to `Added`.

File: zcash_client_backend/src/data_api.rs [hunk 14]
```diff
@@ -248,21 +248,27 @@ pub struct SentTransaction<'a> {
...
 }
 
-pub struct SentTransactionOutput<'a> {
+#[derive(Debug, Clone)]
+pub enum Recipient {
+    Address(RecipientAddress),
+    InternalAccount(AccountId),
+}
+
+pub struct SentTransactionOutput {
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply zcash_client_backend/src/data_api.rs:273]: This type is new.

File: zcash_client_sqlite/src/prepared.rs (hunk 11 above)
Comment (str4d) [reply zcash_client_sqlite/src/prepared.rs:351]: It _was_ a bug at the time, but it looks like the most recent commit fixes it.

File: zcash_client_sqlite/src/wallet.rs (hunk 5 above)
Comment (str4d) [reply zcash_client_sqlite/src/wallet.rs]: It _was_ invalid at the time (it's not possible to have a transparent recipient in the Sapling pool), but it looks like the most recent commit fixes it.

File: zcash_client_sqlite/src/prepared.rs [hunk 15]
```diff
@@ -346,6 +354,103 @@ impl<'a, P> DataConnStmtCache<'a, P> {
 }
 
 impl<'a, P: consensus::Parameters> DataConnStmtCache<'a, P> {
+    /// Inserts a sent note into the wallet database.
+    ///
+    /// `output_index` is the index within the transaction that contains the recipient output:
+    ///
+    /// - If `to` is a Sapling address, this is an index into the Sapling outputs of the
+    ///   transaction.
```
Comment (str4d) [inline zcash_client_sqlite/src/prepared.rs:364]: This does not mention UAs.

File: zcash_client_sqlite/CHANGELOG.md [hunk 16]
```diff
@@ -131,7 +134,6 @@ and this library adheres to Rust's notion of
     - `insert_witness`
     - `prune_witnesses`
     - `update_expired_notes`
-    - `put_sent_note`
     - `put_sent_utxo`
     - `insert_sent_note`
     - `insert_sent_utxo`
```
Comment (str4d) [inline zcash_client_sqlite/CHANGELOG.md]:
```suggestion
```

**Review by str4d — CHANGES_REQUESTED** [review]
Reviewed a8df4c2587682166575aaa7a167032ab9b815db1

**Review by str4d — APPROVED** [review]
utACK e666e692301f7fde2e873d04c84f95e3c24a24a0.
